### PR TITLE
storage: validate disk size range of custom disk offering when resize volume

### LINKED ISSUE #6774: Max size for custom disk offering is checked for disk creation but not for resizing
##### ISSUE TYPE

 * Bug Report

##### COMPONENT NAME
server

##### CLOUDSTACK VERSION

All versions

##### CONFIGURATION
N/A

##### OS / ENVIRONMENT
N/A

##### SUMMARY
The maximum size for custom disk offering is checked for disk creation but not for resizing.
Users cannot create volumes larger than `CustomDiskOfferingMaxSize` but can resize their volumes afterwards to larger sizes up to the allowed quota.

In the code, `CustomDiskOfferingMaxSize` is checked in [`allocVolume`](https://github.com/apache/cloudstack/blob/8ab8fdc1db4fed4d906c69618980fda9ef710e6a/server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java#L718) but not in [`resizeVolume`](https://github.com/apache/cloudstack/blob/8ab8fdc1db4fed4d906c69618980fda9ef710e6a/server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java#L1020).

##### STEPS TO REPRODUCE
- Make sure that there is a custom disk offering
- Try to create a volume with a custom disk offering and a size larger than `CustomDiskOfferingMaxSize` but lower than other quotas.
- Try to resize a volume with a custom disk offering and a size larger than `CustomDiskOfferingMaxSize` but lower than other quotas.

##### EXPECTED RESULTS
Both creation and resizing should fail.

##### ACTUAL RESULTS
Disk creation fails but resizing succeeds.

### PR DESCRIPTION
**Pull request by weizhouapache**
### Description

This PR fixes #6774

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [ ] Major
- [ ] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [ ] Minor
- [ ] Trivial

### Screenshots (if appropriate):

set global setting
- custom.diskoffering.size.max = 20
- custom.diskoffering.size.min = 10

resize volume
- size too big
image

- size too small
image

### CODE REVIEW

**Review by DaanHoogland — APPROVED** [review]
clgtm, guess it is obvious

File: server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java [hunk 1]
```diff
@@ -872,6 +867,16 @@ public VolumeVO allocVolume(CreateVolumeCmd cmd) throws ResourceAllocationExcept
                 _uuidMgr.generateUuid(Volume.class, cmd.getCustomId()), details);
     }
 
+    @Override
+    public void validateCustomDiskOfferingSizeRange(Long sizeInGB) {
+        Long customDiskOfferingMaxSize = VolumeOrchestrationService.CustomDiskOfferingMaxSize.value();
+        Long customDiskOfferingMinSize = VolumeOrchestrationService.CustomDiskOfferingMinSize.value();
+
+        if ((sizeInGB < customDiskOfferingMinSize) || (sizeInGB > customDiskOfferingMaxSize)) {
+            throw new InvalidParameterValueException("Volume size: " + sizeInGB + "GB is out of allowed range. Max: " + customDiskOfferingMaxSize + " Min:" + customDiskOfferingMinSize);
```
Comment (stephankruggg) [inline server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java:876]:
```suggestion
            throw new InvalidParameterValueException(String.format("Volume size:  %sGB is out of allowed range. Min:  %s. Max: %s." sizeInGB, customDiskOfferingMinSize, customDiskOfferingMaxSize));
```

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test matrix

File: server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java (hunk 1 above)
Comment (DaanHoogland) [reply server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java:876]: @weizhouapache , will you apply this suggestion?

Comment (weizhouapache) [reply server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java]: @DaanHoogland done

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test keepEnv

**Review by stephankruggg — APPROVED** [review]

**Review by stephankruggg — APPROVED** [review]
CLGTM

**Review by rohityadavcloud — APPROVED** [review]

**Pull request merged**